scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -79,6 +77,8 @@ class DynamicTLSPeerGroup[M](val config: Config)(
 
   private lazy val serverBind: ChannelFuture = serverBootstrap.bind(config.bindAddress)
 
+  // FIXME We could probably delete this funcion as it all peer groups it just start a server and there is already `server` function
+  // it could return Task[Observable[ServerEvent[PeerInfo, M]]]
   override def initialize(): Task[Unit] =
```

Comment:
I think it would be even better to remove `initialize()` and `shutdown()` from the `PeerGroup` itself and instead use a [Resource](https://typelevel.org/cats-effect/api/cats/effect/Resource.html) to wrap creation of a peer group instance and encapsulate the cleanup as well:

```scala
object DynamicTLSPeerGroup {
  def apply(config: Config)(
    implicit codec: StreamCodec[M],
    scheduler: Scheduler
  ): Resource[Task, DynamicTLSPeerGroup[M]] = 
  Resource.make(
    for {
      pg <- Task.eval(new DynamicTLSPeerGroup(config))
      _   <- pg.initialize()
    } yield pg
  )(_.shutdown())
}
```

This would make sure that users of the `PeerGroup` trait itself can't just call `.shutdown()` whenever they want.

Reply:
I am not sure about it to be honest, as a user of library i would like to close my resources as i want. I can of course use bracket and compose it nicely with the rest of the code, but maybe i need to shutdown it along with actorsystem in some random place.

Reply:
You can do that by calling [Resource.allocated](https://typelevel.org/cats-effect/api/cats/effect/Resource.html#allocated[G[x]%3E:F[x],B%3E:A](implicitF:cats.effect.Bracket[[%CE%B1$7$]G[%CE%B1$7$],Throwable]):G[(B,G[Unit])]) which gives you back the method to call if you want to free your resource explicitly. 

My suggestion is that the ActorSystem and in fact everything that needs closing should be refactored into a `Resource`.